### ISSUE TITLE
Add details on authentication / permissions used during upload

### ISSUE THREAD

**Issue by letmaik** [actions/upload-artifact#197]
With GitHub per-job tokens moving to fine-grained permissions I was wondering how this action authenticates in order to upload artifacts. Does it use the token? If so, which permissions are needed? If not, how is authentication handled instead?

**Comment by letmaik** [comment]
After some searching through GitHub I found out that this action relies on a special [undocumented token](https://github.com/actions/toolkit/blob/e9c6ee99a59ec34586d4374b68efe5cabfcee9db/packages/artifact/src/internal/config-variables.ts#L34) that is made available to actions as environment variable and is unrelated to GITHUB_TOKEN. Side note: this token is not made available to `run:` scripts, only to actions.

I created a small [demo](https://github.com/letmaik/upload-artifact/blob/main/.github/workflows/ci.yml) that [shows](https://github.com/letmaik/upload-artifact/actions/runs/811483742) how with the code of the upload-artifact action someone can create a seemingly innocent action that overwrites previously uploaded artifacts. This is not great if publishing of artifacts to package repositories is fully automated.

One possible solution would be if artifacts could be marked as immutable so that they can't be overridden, and maybe that should be the default.

**Comment by QuintinWillison** [comment]
I've stumbled across this issue as I'm currently exploring the implications for our workflow files of moving our org default access level from 'permissive' to 'restricted' (see [Permissions for the `GITHUB_TOKEN`](https://docs.github.com/en/actions/reference/authentication-in-a-workflow#permissions-for-the-github_token) and [workflow syntax: `permissions`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions)).

The greater focus on finer grained permissions scope is a welcome improvement to the GitHub Actions ecosystem, however I think the developer experience would be greatly enhanced were GitHub to take the initiative to explicitly document permissions required (in the same terms as referred to in workflow files) in the actions published under [this org](https://github.com/actions).

At the time of writing this I'm yet to discover whether invoking this action with the default restricted permissions will fail my workflows or not. I could, perhaps, spend some time exploring the multiple layers of source code orchestrated to achieve what this action does - but, honestly, I don't have the time for that. So, as is so often the case, I'm going to resort to 'suck and see' - i.e. make the change and see if stuff breaks. Not ideal.

**Comment by laurentsimon** [comment]
agreed there's a lack of support. Once you understand how permissions work, you can make an informed decision. But it takes a little getting used to :-)

In the meantime, try https://github.com/naveensrinivasan/ghactionspermissions, written by someone on my team https://github.com/ossf/scorecard
FYI @naveensrinivasan

**Comment by naveensrinivasan** [comment]
> agreed there's a lack of support. Once you understand how permissions work, you can make an informed decision. But it takes a little getting used to :-)
> 
> In the meantime, try https://github.com/naveensrinivasan/ghactionspermissions, written by someone on my team https://github.com/ossf/scorecard
> FYI @naveensrinivasan

Don't use that tool. Sorry, it is not the right approach. Here is the discussion https://github.com/github/codeql-action/pull/689#issuecomment-919402768
